### ISSUE TITLE
Switch back to doubles

### ISSUE THREAD

**Issue by N-Wouda** [PyVRP/PyVRP#623]
#### Is your feature request related to a problem? Please describe

Modelling fractional costs (e.g., 21ct/km, 19.95 EUR/h) gets out of hand quick with all integer data. We have already seen that before, and this was one of the main motivators for 64-bit integers in #468. However, in #550 we need to do weird things to account for this again. Specifying fractional costs is just fundamentally awkward in integers.

Additionally, integers overflow. This has already caused at least one bug (e.g., #590) and makes the code more brittle because we need to work around the issue (e.g., #599, #586). Further, the lack of builtin support for infinity means we need to use a large `MAX_VALUE` constant instead to avoid the overflow problem - with limited success (e.g., fails in #617, #550).

Floats have their own problems. For example, equality checks between floats needs some tolerance that is [a bit iffy](https://stackoverflow.com/q/17333/4316405).

#### Describe the solution you'd like

Switch costs, distances, durations etc. back over to float or double, rather than the integral values we impose now. I want to revert #518, re-instate #47, and probably remove integer support instead.

#### Describe alternatives you have considered

This will remain awkward in any other way, so there aren't really any. 64-bit integers give us a lot of flexibility already, but without hardware support for infinity (or NaN) and always the risk of overflows it does not seem to be worth it.

#### Additional context

Most of the double vs int discussions previously focused on performance (ints were a bit faster). Most of that was likely due to us using 32-bit integers, which are half the size of a double. We now use 64-bit integers since we need the extra space for modelling (see #468), so that should essentially remove the performance issue.

**Comment by N-Wouda** [comment]
A question is whether we need `double` (=float64) or `float` (=float32). 32 bits might be enough, but I do not know that for sure. We could experiment to find out?

**Comment by leonlan** [comment]
I'm in favor of having floats because it improves user experience by not having to scale all the time. I notice this in the benchmarks, especially since we now have unit distance costs (#550). This requires scaling twice and often results in overflows.

What are the other downsides of switching back to floats? I recall from somewhere that we get about 25% less iterations. That's a significant number of iterations, but we can probably get that performance back from somewhere else.

**Comment by N-Wouda** [comment]
> What are the other downsides of switching back to floats? I recall from somewhere that we get about 25% less iterations. That's a significant number of iterations, but we can probably get that performance back from somewhere else.

IIRC that was a pretty old experiment. I think most of that was due to the switch from 32-bit integers to 64-bit floats, which doubled the space requirements and limited the data we could keep in cache. But we already switched to 64-bit integers, which means we paid most of that expense already.

We should probably check to make sure, but I wouldn't be surprised if the switch is now basically performance neutral.

**Comment by leonlan** [comment]
> We should probably check to make sure, but I wouldn't be surprised if the switch is now basically performance neutral.

Checking this with commit https://github.com/PyVRP/PyVRP/commit/0d8d428c507ec7f397776124bdfe662c740067e8.

**Comment by leonlan** [comment]
With 5 seeds, 30 min per instance:
- Main with ints: 26.1K average iterations
- https://github.com/PyVRP/PyVRP/commit/5dad99ad6a14118b5b5b0531cdbb6caa2a63ecd1 with doubles: 21.8K average iterations

So that's about 20% fewer iterations. Could it be due to subtle differences in how moves are evaluated?

**Comment by N-Wouda** [comment]
@leonlan maybe? I don't know enough about this to be sure.

**Comment by N-Wouda** [comment]
> A question is whether we need `double` (=float64) or `float` (=float32). 32 bits might be enough, but I do not know that for sure. We could experiment to find out?

Ah, I already wrote what I (again) wanted to ask now. We should really check this. @leonlan is this something you could run an experiment for? Perhaps again on the same branch as https://github.com/PyVRP/PyVRP/commit/5dad99ad6a14118b5b5b0531cdbb6caa2a63ecd1 was ran on, so we can compare against the results from last month.

**Comment by leonlan** [comment]
GH1000 instances, with 5 seeds, 30 min per instance:
- https://github.com/PyVRP/PyVRP/commit/9be2b6422230e4f3781b0d10388274cb1dda10db with `int64_t`: 26.1K average iterations
- https://github.com/PyVRP/PyVRP/commit/5dad99ad6a14118b5b5b0531cdbb6caa2a63ecd1 with `double`: 21.8K average iterations
- https://github.com/PyVRP/PyVRP/commit/8938e883320825f4b4f6949ba49a2d534c46efb7 with `float`: 27.1K average iterations

**Comment by N-Wouda** [comment]
That looks pretty good to me. Undoubtedly we can optimise the implementation a bit further but for now I am already pretty happy with this result. So I guess the bigger question is whether `float` is sufficient for what we do here. A `float` can accurately represent up to 7 decimal digits. E.g.,
- 1234.567
- 123.4567
- 1234567
- 123456.7

can all be represented accurately. Anything more than this incurs some error because it cannot be represented all that accurately. I _think_ this precision should suffice, but I do not know for sure.

**Comment by N-Wouda** [comment]
This is interesting enough to put on the list for 0.11.0, and resolve by then. It's too much for 0.10.0, which I'd like to release pretty soon (after I complete #645, which I'll work on after lunch).

**Comment by N-Wouda** [comment]
Another bunch of issues related to integers not being floats: #683, #678, #691. And that's just the past week. I think we should bite the bullet here, performance loss or not.

**Comment by N-Wouda** [comment]
I had a go at this in #703 just yet. It's (much) harder than I'd have liked. We need to account for all sorts of small/annoying issues like:
- float approximate equality working with both small and large floats.
- performance taking a hit because equality is checked _everywhere_.
- what are good default values? inf, +max?

**Comment by N-Wouda** [comment]
I don't think this is worth it. There's a lot of benefit to having floating-point values, but the downsides make me think we should not pursue this further. I have yet to find a real-world case I could not model sufficiently well using integers.

**Comment by N-Wouda** [comment]
Re-opening this. Floats do two things I like:
- They fix our overflow worries. Sufficiently documented earlier in this issue.
- For PWL functions, they let us specify a wider range of possible functions, without having to carefully scale all data. Additionally, they model jumps better because we can make the segment endpoints inclusive. See #1052 for more details.

**Comment by 0nyr** [comment]
One data point from [another codebase (Kayros)](https://github.com/0nyr/kayros) on the 2026-03-18 reopening: our TD-VRP solver (Kayros, exact + anytime over continuous PWL arrival-time functions) went all-in on doubles with exact IEEE-754 comparisons (no epsilons in the engine, `-ffp-contract=off`, bit-reproducible across two toolchains), refereeing every value against an independent Python checker. That is viable for us because an exact labeling solver reprices accepted states anyway. The approximate-equality cost inside a delta-evaluation metaheuristic is real and we do not pay it in the same place, so it is a neighboring data point rather than a counterexample to your concerns. The jump / inclusive-endpoint semantics were the hard part for us too: we ended up carrying verticals as tagged first-class objects through compose/min/max after getting them wrong once (and retracting 160 published optimality certificates as a result). 

Details are in the technical report: https://arxiv.org/abs/2607.23116.

**Comment by N-Wouda** [comment]
Thanks @0nyr for your account. That reminded me of this issue still/again being open. 

We have flip-flopped on this quite a bit the past years, but realistically, I don't expect we're actually going to switch in the future. PyVRP works very well with integers. The exact integer representation has a lot of benefit, while the downsides are manageable.
